### PR TITLE
Started P2P rework

### PR DESCRIPTION
**Pull request by yueh**
Rebase of #2692 onto rv5-1.12

### CODE REVIEW

File: src/main/java/appeng/me/cache/P2PCache.java [hunk 1]
```diff
@@ -20,8 +20,10 @@
 
 
 import java.util.HashMap;
+import java.util.Random;
 
 import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Maps;
```
Comment (orod-org) [inline src/main/java/appeng/me/cache/P2PCache.java]: MINOR Remove this unused import 'com.google.common.collect.Maps'. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/appeng/me/cache/P2PCache.java [hunk 2]
```diff
@@ -34,22 +36,26 @@
 import appeng.api.networking.events.MENetworkEventSubscribe;
 import appeng.api.networking.events.MENetworkPowerStatusChange;
 import appeng.api.networking.ticking.ITickManager;
+import appeng.core.AELog;
 import appeng.me.cache.helpers.TunnelCollection;
 import appeng.parts.p2p.PartP2PTunnel;
 import appeng.parts.p2p.PartP2PTunnelME;
+import appeng.util.Platform;
```
Comment (orod-org) [inline src/main/java/appeng/me/cache/P2PCache.java]: MINOR Remove this unused import 'appeng.util.Platform'. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by orod-org** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [PartP2PTunnel.java#L213](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/ad04fabf71261d1d53bd3c9f9f0fe0af2f269dc6/src/main/java/appeng/parts/p2p/PartP2PTunnel.java#L213): Remove this expression which always evaluates to "true" [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2589)
1. ![MAJOR][MAJOR] [Platform.java#L690](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/ad04fabf71261d1d53bd3c9f9f0fe0af2f269dc6/src/main/java/appeng/util/Platform.java#L690): Remove this useless assignment to local variable "itemStack". [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Platform.java#L735](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/ad04fabf71261d1d53bd3c9f9f0fe0af2f269dc6/src/main/java/appeng/util/Platform.java#L735): Remove this useless assignment to local variable "itemStack". [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Platform.java#L1813](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/ad04fabf71261d1d53bd3c9f9f0fe0af2f269dc6/src/main/java/appeng/util/Platform.java#L1813): This block of commented-out lines of code should be removed. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MINOR][MINOR] [Platform.java#L467](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/ad04fabf71261d1d53bd3c9f9f0fe0af2f269dc6/src/main/java/appeng/util/Platform.java#L467): Remove this use of "getDrops"; it is deprecated. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/appeng/util/helpers/P2PHelper.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+public class P2PHelper
+{
+
+	public AEColor[] toColours( short frequency )
+	{
+		final AEColor[] colours = new AEColor[4];
+
+		for( int i = 0; i < 4; i++ )
+		{
+			int twoBits = ( frequency >> 4 * ( 3 - i ) ) & 0xF;
```
Comment (tyra314) [inline src/main/java/appeng/util/helpers/P2PHelper.java]: Actually, this looks more like 4 bit. ` & 0x3` would give two bit

File: src/main/java/appeng/spatial/StorageHelper.java [hunk 4]
```diff
@@ -139,6 +139,8 @@ private Entity teleportEntity( Entity entity, final TelDestination link )
 				passanger.startRiding( entity, true );
 			}
 		}
+		
+		entity.world.updateEntity( entity );
```
Comment (tyra314) [inline src/main/java/appeng/spatial/StorageHelper.java]: Spatial? This looks unrelated to P2P

Comment (yueh) [inline src/main/java/appeng/spatial/StorageHelper.java]: Rebase artifact. But might be related to #2654?

File: src/main/java/appeng/util/helpers/P2PHelper.java (hunk 3 above)
Comment (yueh) [inline src/main/java/appeng/util/helpers/P2PHelper.java]: I cannot recall how it was named, but it needs to be a nibble. So `0xF` is correct.

**Comment by yueh** [comment]
Also related to  #2674

**Comment by orod-org** [comment]
SonarQube analysis reported 5 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [PartP2PTunnel.java#L213](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/c8e25085798563293f9c0f2878e8a3683821b71b/src/main/java/appeng/parts/p2p/PartP2PTunnel.java#L213): Remove this expression which always evaluates to "true" [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2589)
1. ![MAJOR][MAJOR] [Platform.java#L690](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/c8e25085798563293f9c0f2878e8a3683821b71b/src/main/java/appeng/util/Platform.java#L690): Remove this useless assignment to local variable "itemStack". [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Platform.java#L735](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/c8e25085798563293f9c0f2878e8a3683821b71b/src/main/java/appeng/util/Platform.java#L735): Remove this useless assignment to local variable "itemStack". [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Platform.java#L1813](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/c8e25085798563293f9c0f2878e8a3683821b71b/src/main/java/appeng/util/Platform.java#L1813): This block of commented-out lines of code should be removed. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MINOR][MINOR] [Platform.java#L467](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/c8e25085798563293f9c0f2878e8a3683821b71b/src/main/java/appeng/util/Platform.java#L467): Remove this use of "getDrops"; it is deprecated. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/appeng/spatial/StorageHelper.java (hunk 4 above)
Comment (tyra314) [inline src/main/java/appeng/spatial/StorageHelper.java]: It might be related, but it doesn't fix that issue just like that. I tried that yesterday.

File: src/main/java/appeng/util/helpers/P2PHelper.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+		int t = 0;
+
+		for( int i = 0; i < 4; i++ )
+		{
+			int code = colors[3 - i].ordinal() << 4 * i;
+
+			t |= code;
+		}
+
+		return (short) ( t & 0xFFFF );
```
Comment (tyra314) [inline src/main/java/appeng/util/helpers/P2PHelper.java]: As `ordinal()` returns values between 0 and 16 (including), `t`can only have bits set outside of 0xFFFF, if `colors[0] == AEColor.TRANSPARENT`. If `colors[1]` to `colors[3]` contain `AEColor.TRANSPARENT`, the higher nibble is potentially messed up, but it will never go beyond the 0xFFFF bitmask. If I understand the idea correctly, we assume that colors will only ever contain one of the 16 "real" colors. In this case, I would argue to remove the bitmask and - if you think it is necessary - check all entries of `colors` against `AEColor.TRANSPARENT`.

File: src/main/java/appeng/util/helpers/P2PHelper.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import appeng.api.util.AEColor;
+
+
+public class P2PHelper
+{
+
+	public AEColor[] toColors( short frequency )
+	{
+		final AEColor[] colours = new AEColor[4];
```
Comment (tyra314) [inline src/main/java/appeng/util/helpers/P2PHelper.java]: You missed one `colours` here :)

File: src/main/java/appeng/util/helpers/P2PHelper.java (hunk 5 above)
Comment (yueh) [inline src/main/java/appeng/util/helpers/P2PHelper.java]: `AEColor.TRANSPARENT` is not a color. It is the absence of any color (or any depending on the use case). The only reason it is part of the enum, is that wrapping everything into a null check and then having to fall back to default values is tedious. Similar to why we use `AEPartLocation` and not `EnumFacing`.

Sure I could safeguard it, but it is never intended to be used externally. Or probably not even internally. The sole reason I added it, is because I can test the `toColors` more easily.

**Comment by orod-org** [comment]
SonarQube analysis reported 5 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [PartP2PTunnel.java#L213](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b67d0e76c45ba217651747e1064e18e36e3253a8/src/main/java/appeng/parts/p2p/PartP2PTunnel.java#L213): Remove this expression which always evaluates to "true" [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2589)
1. ![MAJOR][MAJOR] [Platform.java#L690](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b67d0e76c45ba217651747e1064e18e36e3253a8/src/main/java/appeng/util/Platform.java#L690): Remove this useless assignment to local variable "itemStack". [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Platform.java#L735](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b67d0e76c45ba217651747e1064e18e36e3253a8/src/main/java/appeng/util/Platform.java#L735): Remove this useless assignment to local variable "itemStack". [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Platform.java#L1813](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b67d0e76c45ba217651747e1064e18e36e3253a8/src/main/java/appeng/util/Platform.java#L1813): This block of commented-out lines of code should be removed. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MINOR][MINOR] [Platform.java#L467](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/b67d0e76c45ba217651747e1064e18e36e3253a8/src/main/java/appeng/util/Platform.java#L467): Remove this use of "getDrops"; it is deprecated. [rule](https://sonar.orod.org/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'
